src/Npgsql/NpgsqlDataReader.cs

```diff
@@ -2167,7 +2167,14 @@ internal void UnbindIfNecessary()
             // We have to unbind the reader from the connector, otherwise there could be a concurency issues
             // See #3126 and #3290
             if (State != ReaderState.Disposed)
-                Connector.DataReader = new NpgsqlDataReader(Connector);
+            {
+                var previousReader = Connector.UnboundDataReader;
+                if (previousReader is not null && previousReader.State == ReaderState.Disposed)
```

Comment:
Go further, use pattern matching!

```csharp
Connector.DataReader = Connector.UnboundDataReader is { State: ReaderState.Disposed } previousReader
    ? previousReader
    : new NpgsqlDataReader(Connector);
```

It's much shorter and laconic. Plus no one used it before here, be first!

Reply:
Yep, nice syntax!